core/store/src/trie/trie_storage.rs

```diff
@@ -560,10 +560,16 @@ impl TrieStorage for TrieCachingStorage {
                                 // therefore blocking read will usually not return empty unless there
                                 // was a storage error. Or in the case of forks and parallel chunk
                                 // processing where one chunk cleans up prefetched data from the other.
-                                // In any case, we can try again from the main thread.
+                                // So first we need to check if the data was inserted to shard_cache
+                                // by the main thread from another fork and only if that fails then
+                                // fetch the data from the DB.
                                 None => {
-                                    self.metrics.prefetch_retry.inc();
-                                    self.read_from_db(hash)?
+                                    if let Some(value) = self.shard_cache.get(hash) {
+                                        value
```

Comment:
@jakmeier do you think it is worth introducing another metric for this case, something like:
```
self.metrics.prefetch_shared_cache_after_pending.inc();
```
also suggestions for a better name are welcome :)

Reply:
Hm, yeah I think we can add it. It's exactly these weird corner cases that are hard to debug, having insight with metrics might turn out useful.

For the name, do you mean `shard_cache` or `shared_cache`? I think `prefetch_shard_cache_after_pending` makes more sense to me.

But actually, how about `prefetch_pending_conflict`, or `prefetch_pending_conflict_hit`, or `prefetch_main_thread_conflict` or even just `prefetch_conflict`?

My thinking is: The request was pending but instead of the expected prefetch-hit afterwards, there was a conflict with "another main thread" that also lead to a hit, but the hit was unexpectedly in shard cache. You name also work though, these are just some suggestions.

Reply:
it meant to be `shard_cache`, sorry for the typo

`prefetch_conflict` definitely makes to me, let's use it